File: MATTHEW/leetcode/AIcoding/1.Gemini-2.5/rust_solutions/1930-unique-length-3-palindromic-subsequences.rs
use std::collections::HashSet;

impl Solution {
    pub fn count_palindromic_subsequences(s: String) -> i32 {
        let mut ans = 0;

        for i in 0..26 {
            let c_outer = (b'a' + i) as char;

            let first_idx = s.find(c_outer);
            let last_idx = s.rfind(c_outer);

            if let (Some(f_idx), Some(l_idx)) = (first_idx, last_idx) {
                if f_idx < l_idx {
                    let mut unique_middle_chars = HashSet::new();

                    // Get the slice of the string between the first and last occurrences of c_outer
                    // If f_idx + 1 >= l_idx, this slice will be empty or invalid,
                    // and s.get() will return None or an empty slice, which is handled correctly.
                    if let Some(middle_slice) = s.get(f_idx + 1..l_idx) {
                        for c in middle_slice.chars() {
                            unique_middle_chars.insert(c);
                        }
                    }
                    ans += unique_middle_chars.len();
                }
            }
        }
        ans as i32
    }
}

File: leetcode/AIcoding/1.Gemini-2.5/java_solutions/1930-unique-length-3-palindromic-subsequences.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int countPalindromicSubsequences(String s) {
        int totalUniquePalindromes = 0;

        for (char c = 'a'; c <= 'z'; c++) {
            int firstIdx = -1;
            int lastIdx = -1;

            for (int i = 0; i < s.length(); i++) {
                if (s.charAt(i) == c) {
                    firstIdx = i;
                    break;
                }
            }

            if (firstIdx == -1) {
                continue;
            }

            for (int i = s.length() - 1; i >= 0; i--) {
                if (s.charAt(i) == c) {
                    lastIdx = i;
                    break;
                }
            }

            if (firstIdx == lastIdx) {
                continue;
            }

            Set<Character> uniqueMiddleChars = new HashSet<>();
            for (int i = firstIdx + 1; i < lastIdx; i++) {
                uniqueMiddleChars.add(s.charAt(i));
            }

            totalUniquePalindromes += uniqueMiddleChars.size();
        }

        return totalUniquePalindromes;
    }
}